File: backend/app/schemas/enrichment.py
from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any

from pydantic import BaseModel, ConfigDict


# ---------------------------------------------------------------------------
# Apollo API response models (inbound boundary — shape Apollo returns to us)
# ---------------------------------------------------------------------------

class ApolloPhoneNumber(BaseModel):
    model_config = ConfigDict(extra="ignore")

    raw_number: str | None = None
    sanitized_number: str | None = None
    type: str | None = None  # "work" | "mobile" | "home"


class ApolloEmploymentEntry(BaseModel):
    model_config = ConfigDict(extra="ignore")

    title: str | None = None
    organization_name: str | None = None
    start_date: str | None = None  # "YYYY-MM-DD" or "YYYY-01-01"
    end_date: str | None = None
    current: bool = False


class ApolloOrganization(BaseModel):
    model_config = ConfigDict(extra="ignore")

    name: str | None = None
    website_url: str | None = None
    linkedin_url: str | None = None
    primary_domain: str | None = None       # upsert key → companies.domain
    industry: str | None = None             # → companies.industry
    estimated_num_employees: int | None = None  # → companies.size (bucketed)
    revenue_range: str | None = None        # → companies.revenue
    company_type: str | None = None         # → companies.company_type
    short_description: str | None = None    # → companies.overview
    city: str | None = None
    state: str | None = None
    country: str | None = None


class ApolloPerson(BaseModel):
    model_config = ConfigDict(extra="ignore")

    first_name: str | None = None
    last_name: str | None = None
    linkedin_url: str | None = None
    email: str | None = None
    email_status: str | None = None  # "verified" | "likely valid" | "invalid"
    phone_numbers: list[ApolloPhoneNumber] = []
    title: str | None = None         # current job title
    seniority: str | None = None     # "vp" | "director" | "manager" | "entry"
    city: str | None = None
    state: str | None = None
    country: str | None = None
    employment_history: list[ApolloEmploymentEntry] = []
    organization: ApolloOrganization | None = None


class ApolloPersonResponse(BaseModel):
    model_config = ConfigDict(extra="ignore")

    person: ApolloPerson | None = None


# ---------------------------------------------------------------------------
# Enrichment run models (outbound — what we write to enrichment_runs table)
# ---------------------------------------------------------------------------

class EnrichmentRunType(str, Enum):
    initial = "initial"
    scheduled = "scheduled"
    manual = "manual"


class EnrichmentRunStatus(str, Enum):
    success = "success"
    failed = "failed"
    partial = "partial"


class EnrichmentRunCreate(BaseModel):
    member_id: str
    run_type: EnrichmentRunType
    status: EnrichmentRunStatus
    fields_updated: dict[str, Any] = {}
    fields_skipped: dict[str, Any] = {}
    error_message: str | None = None


class EnrichmentRunResponse(EnrichmentRunCreate):
    model_config = ConfigDict(from_attributes=True)

    id: str
    ran_at: datetime


# ---------------------------------------------------------------------------
# API surface models (trigger request / response)
# ---------------------------------------------------------------------------

class EnrichmentTriggerRequest(BaseModel):
    member_id: str
    run_type: EnrichmentRunType = EnrichmentRunType.manual


class EnrichmentTriggerResponse(BaseModel):
    run_id: str
    status: EnrichmentRunStatus
    fields_updated: dict[str, Any]
    fields_skipped: dict[str, Any]

File: backend/src/schemas/enrichment.ts
import { z } from "zod";

export const ApolloPhoneNumberSchema = z.object({
  raw_number: z.string().nullable().optional(),
  sanitized_number: z.string().nullable().optional(),
  type: z.string().nullable().optional(),
});

export const ApolloEmploymentEntrySchema = z.object({
  title: z.string().nullable().optional(),
  organization_name: z.string().nullable().optional(),
  start_date: z.string().nullable().optional(),
  end_date: z.string().nullable().optional(),
  current: z.boolean().default(false),
});
export type ApolloEmploymentEntry = z.infer<typeof ApolloEmploymentEntrySchema>;

export const ApolloOrganizationSchema = z.object({
  name: z.string().nullable().optional(),
  website_url: z.string().nullable().optional(),
  linkedin_url: z.string().nullable().optional(),
  primary_domain: z.string().nullable().optional(),
  industry: z.string().nullable().optional(),
  estimated_num_employees: z.coerce.number().int().nullable().optional(),
  revenue_range: z.string().nullable().optional(),
  company_type: z.string().nullable().optional(),
  short_description: z.string().nullable().optional(),
  city: z.string().nullable().optional(),
  state: z.string().nullable().optional(),
  country: z.string().nullable().optional(),
});
export type ApolloOrganization = z.infer<typeof ApolloOrganizationSchema>;

export const ApolloPersonSchema = z.object({
  first_name: z.string().nullable().optional(),
  last_name: z.string().nullable().optional(),
  linkedin_url: z.string().nullable().optional(),
  email: z.string().nullable().optional(),
  email_status: z.string().nullable().optional(),
  phone_numbers: z.array(ApolloPhoneNumberSchema).default([]),
  title: z.string().nullable().optional(),
  seniority: z.string().nullable().optional(),
  city: z.string().nullable().optional(),
  state: z.string().nullable().optional(),
  country: z.string().nullable().optional(),
  employment_history: z.array(ApolloEmploymentEntrySchema).default([]),
  organization: ApolloOrganizationSchema.nullable().optional(),
});
export type ApolloPerson = z.infer<typeof ApolloPersonSchema>;

export const ApolloPersonResponseSchema = z.object({
  person: ApolloPersonSchema.nullable().optional(),
});
export type ApolloPersonResponse = z.infer<typeof ApolloPersonResponseSchema>;

export const EnrichmentRunTypeSchema = z.enum(["initial", "scheduled", "manual"]);
export type EnrichmentRunType = z.infer<typeof EnrichmentRunTypeSchema>;

export const EnrichmentRunStatusSchema = z.enum(["success", "failed", "partial"]);
export type EnrichmentRunStatus = z.infer<typeof EnrichmentRunStatusSchema>;

export const EnrichmentRunCreateSchema = z.object({
  member_id: z.string(),
  run_type: EnrichmentRunTypeSchema,
  status: EnrichmentRunStatusSchema,
  fields_updated: z.record(z.unknown()).default({}),
  fields_skipped: z.record(z.unknown()).default({}),
  error_message: z.string().nullable().optional(),
});
export type EnrichmentRunCreate = z.infer<typeof EnrichmentRunCreateSchema>;

export const EnrichmentRunResponseSchema = EnrichmentRunCreateSchema.extend({
  id: z.string(),
  ran_at: z.string(),
});
export type EnrichmentRunResponse = z.infer<typeof EnrichmentRunResponseSchema>;

export const EnrichmentTriggerRequestSchema = z.object({
  member_id: z.string(),
  run_type: EnrichmentRunTypeSchema.default("manual"),
});
export type EnrichmentTriggerRequest = z.infer<typeof EnrichmentTriggerRequestSchema>;

export const EnrichmentTriggerResponseSchema = z.object({
  run_id: z.string(),
  status: EnrichmentRunStatusSchema,
  fields_updated: z.record(z.unknown()),
  fields_skipped: z.record(z.unknown()),
});
export type EnrichmentTriggerResponse = z.infer<typeof EnrichmentTriggerResponseSchema>;
